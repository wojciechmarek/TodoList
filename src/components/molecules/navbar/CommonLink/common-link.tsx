import { CommonLinkProps } from './common-link.interface';
import {
  CommonLinkContainer,
  CommonLinkIcon,
  CommonLinkText,
  CommonLinkStyle,
} from './common-link.styled';

export const CommonLink = ({
  icon,
  text,
  url,
  color,
  expanded,
  ...rest
}: CommonLinkProps) => {
  return (
    <CommonLinkContainer {...rest} color={color}>
      <CommonLinkStyle href={url} target="_blank" rel="noreferrer">
        <CommonLinkIcon icon={icon} />
        <CommonLinkText expanded={expanded} text={text} />
      </CommonLinkStyle>
    </CommonLinkContainer>
  );
};
