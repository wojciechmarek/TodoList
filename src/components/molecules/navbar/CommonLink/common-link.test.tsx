import { configure, render } from '@testing-library/react';
import { CommonLinkProps } from './common-link.interface';
import { CommonLink } from './common-link';
import '@testing-library/jest-dom';
import { MemoryRouter } from 'react-router-dom';

describe('CommonLink', () => {
  let commonLinkProps: CommonLinkProps;

  beforeEach(() => {
    configure({ throwSuggestions: true });

    commonLinkProps = {
      text: 'text',
      color: '#FFF',
      expanded: false,
      icon: <span>A</span>,
      url: 'https://via.placeholder.com/150',
    };
  });

  it('should render', () => {
    const { baseElement } = render(
      <MemoryRouter>
        <CommonLink {...commonLinkProps} />
      </MemoryRouter>
    );
    expect(baseElement).toBeTruthy();
  });

  describe('when the common link is rendered', () => {
    it('should render the text', () => {
      const { getByText } = render(
        <MemoryRouter>
          <CommonLink {...commonLinkProps} />
        </MemoryRouter>
      );
      const text = getByText(/text/i);
      expect(text).toBeTruthy();
    });

    it('should render the icon', () => {
      const { getByText } = render(
        <MemoryRouter>
          <CommonLink {...commonLinkProps} />
        </MemoryRouter>
      );
      const icon = getByText(/A/i);
      expect(icon).toBeTruthy();
    });

    it('should render the url', () => {
      const { getByRole } = render(
        <MemoryRouter>
          <CommonLink {...commonLinkProps} />
        </MemoryRouter>
      );
      const link = getByRole('link');
      expect(link).toBeTruthy();
    });
  });

  describe('when the common link is clicked', () => {
    it('should call the onClick function', () => {
      const { getByRole } = render(
        <MemoryRouter>
          <CommonLink {...commonLinkProps} />
        </MemoryRouter>
      );
      const link = getByRole('link');
      link.click();
      expect(link).toBeTruthy();

      // WE SHOULD NOT TEST DOM EVENTS IN JEST (like click on Link)
      // https://stackoverflow.com/a/58907587

      // TO DO IT - TEST THIS IN CYPRESS
    });
  });
});
