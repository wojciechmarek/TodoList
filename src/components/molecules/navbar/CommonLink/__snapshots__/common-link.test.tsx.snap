// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CommonLink should match the snapshot 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  background-color: #FFF;
  width: 100%;
  -webkit-box-pack: start;
  -ms-flex-pack: start;
  -webkit-justify-content: flex-start;
  justify-content: flex-start;
  -webkit-transition: filter 0.3s ease-in-out;
  transition: filter 0.3s ease-in-out;
  border: var(--standard-border);
  border-radius: var(--standard-border-radius);
  box-shadow: var(--button-box-shadow);
  cursor: pointer;
}

.emotion-0:hover {
  -webkit-filter: brightness(0.9);
  filter: brightness(0.9);
}

.emotion-1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: start;
  -ms-flex-pack: start;
  -webkit-justify-content: start;
  justify-content: start;
  -webkit-text-decoration: none;
  text-decoration: none;
  width: 100%;
}

.emotion-2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  color: var(--button-font-color);
  height: 2.5em;
  width: 2.5em;
}

.emotion-3 {
  height: calc(100% * 0.618);
  width: calc(100% * 0.618);
}

.emotion-4 {
  font-weight: 600;
  color: var(--button-font-color);
  display: none;
}

<li
  className="emotion-0"
  color="#FFF"
>
  <a
    className="emotion-1"
    href="https://via.placeholder.com/150"
  >
    <div
      className="emotion-2"
    >
      <div
        className="emotion-3"
      >
        <span>
          A
        </span>
      </div>
    </div>
    <p
      className="emotion-4"
    >
      text
    </p>
  </a>
</li>
`;
